File: Apps/BankBase/backend/src/main/java/org/d3javu/backend/model/transaction/Transaction.java
package org.d3javu.backend.model.transaction;

import jakarta.persistence.*;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.d3javu.backend.model.account.Account;

import java.time.LocalDateTime;
import java.util.UUID;

@Entity
@Table(name = "transaction")
@Getter
@Setter
@NoArgsConstructor
public class Transaction {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    @ManyToOne(fetch = FetchType.EAGER, optional = false, cascade = CascadeType.ALL)
    @JoinColumn(name = "sender_id")
    private Account sender;

    @ManyToOne(fetch = FetchType.EAGER, optional = false, cascade = CascadeType.ALL)
    @JoinColumn(name = "recipient_id", nullable = false)
    private Account recipient;

    @Column(name = "amount", nullable = false, scale = 2)
    private Double amount;

    @Column(name = "type", nullable = false)
    @Enumerated(EnumType.STRING)
    private TransactionType type;

    @Column(name = "is_completed", nullable = false)
    private boolean isCompleted;

    @Column(name = "commited_at", nullable = false)
    private LocalDateTime commitedAt;

}

File: Apps/BankBase/backend/src/main/scala/model/transaction/Transaction.scala
package model.transaction

import io.getquill.{SchemaMeta, schemaMeta}
import zio.schema.{DeriveSchema, Schema}

import java.time.LocalDateTime
import java.util.UUID

case class Transaction(
                      id: UUID,
                      senderId: Long,
                      recipientId: Option[Long],
                      amount: Double,
                      transactionType: TransactionType,
                      isCompleted: Boolean,
                      commitedAt: LocalDateTime
                      )

object Transaction {
  inline given SchemaMeta[Transaction] = schemaMeta(
    "transaction",
    _.id -> "id",
    _.senderId -> "sender_id",
    _.recipientId -> "recipient_id",
    _.amount -> "amount",
    _.transactionType -> "type",
    _.isCompleted -> "is_completed",
    _.commitedAt -> "commited_at"
  )

  inline given Schema[Transaction] = DeriveSchema.gen[Transaction]
}